### PR TITLE
Adding Docker resource sub-project

### PR DESCRIPTION
**Pull request by trisberg**
- Adding DockerResourceLoader and DockerResource

### CODE REVIEW

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	 * Returns a {@link DockerResource} for the provided location.
+	 *
+	 * @param location the image location. May optionally be preceded by {@value DockerResource#URI_SCHEME}
+	 * followed by a colon, e.g. {@literal docker:springcloud/app-name:tag}
+	 * @return the {@link DockerResource}
+	 */
+	@Override
+	public Resource getResource(String location) {
+		Assert.hasText(location, "image location is required");
+		String image = location.replaceFirst(DockerResource.URI_SCHEME + ":", "");
```
Comment (markfisher) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java]: would "docker://" be allowed as well? if so, the regex could be: `DockerResource.URI_SCHEME + ":\\/*"`

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	 */
+	public DockerResource(String imageName) {
+		this.uri = URI.create(URI_SCHEME + ":" + imageName);
+	}
+
+	/**
+	 * Create a new {@code DockerResource} from a URI
+	 * @param uri a URI
+	 */
+	public DockerResource(URI uri) {
```
Comment (markfisher) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java]: we should probably `Assert.notNull(uri)` also

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+	private String scheme = "docker";
+
+	private URI uri;
+
+	/**
+	 * Create a new {@code DockerResource} from an image name.
+	 * @param imageName the name of the image in a docker registry.
+	 */
+	public DockerResource(String imageName) {
```
Comment (markfisher) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java]: should add an `Assert.hasText(imageName)`

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * Note: {@link #getInputStream()} throws {@code UnsupportedOperationException}.
+ *
+ * @author Thomas Risberg
+ */
+public class DockerResource extends AbstractResource {
+
+	public static String URI_SCHEME = "docker";
+
+	private String scheme = "docker";
```
Comment (markfisher) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java]: might as well use the constant here ;)

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.util.ClassUtils;
+
+/**
+ * A {@link ResourceLoader} that loads {@link DockerResource}s from locations of the format
+ * {@literal docker:<repository>:<tag>} where the value for "repository:tag" conforms to the rules
+ * for referencing Docker images.
+ *
+ * @author Thomas Risberg
+ */
+public class DockerResourceLoader  implements ResourceLoader {
```
Comment (markfisher) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java]: would this eventually include information about one or more particular registries?... and auth credentials if necessary 

assuming we eventually implement the resource's `getInputStream()`, we would need to pass those properties along to the `DockerResource` that is loaded - similar to how `MavenProperties` are passed to a `MavenResource` from the `MavenResourceLoader`... and whereas the Maven implementation uses the Aether library, we would use the Docker Java API

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java (hunk 1 above)
Comment (trisberg) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java]: This is our own scheme so not sure what // would add to it, but I guess we could remove it if provided

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java (hunk 4 above)
Comment (trisberg) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResource.java]: oops

File: spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java (hunk 5 above)
Comment (trisberg) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java]: Yes, we would need to be able to handle credentials, might have to pass them along for deployments. What we have now is bare-bones but gets the work done for now.

Comment (markfisher) [inline spring-cloud-deployer-resource-docker/src/main/java/org/springframework/cloud/deployer/resource/docker/DockerResourceLoader.java]: sounds good

**Comment by markfisher** [comment]
LGTM; merging

**Comment by markfisher** [comment]
squashed and merged as https://github.com/spring-cloud/spring-cloud-deployer/commit/598c1902d75c785a3bfeeecdd8f8f6d9bce06dc5

**Pull request closed without merging**